pkg/causality/internal/node.go

```diff
@@ -212,96 +180,41 @@ func (n *Node) assignTo(workerID int64) bool {
 	defer n.mu.Unlock()
 
 	if n.assignedTo != unassigned {
-		// Already resolved by some other guys.
+		// Already handled by some other guys.
 		return false
 	}
 
 	n.assignedTo = workerID
-	if n.OnResolved != nil {
-		n.OnResolved(workerID)
-		n.OnResolved = nil
-	}
-
-	if n.dependers != nil {
-		// `mu` must be holded during accessing dependers.
-		n.dependers.Ascend(func(node *Node) bool {
-			resolvedDependencies := stdatomic.AddInt32(&node.resolvedDependencies, 1)
-			stdatomic.StoreInt64(&node.resolvedList[resolvedDependencies-1], n.assignedTo)
-			node.maybeResolve(resolvedDependencies, 0)
-			return true
-		})
+	if n.SendToWorker != nil {
+		n.SendToWorker(workerID)
+		n.SendToWorker = nil
 	}
 
 	return true
 }
 
-func (n *Node) maybeResolve(resolvedDependencies, removedDependencies int32) {
-	if workerNum, ok := n.tryResolve(resolvedDependencies, removedDependencies); ok {
-		if workerNum < 0 {
-			panic("Node.tryResolve must return a valid worker ID")
-		}
-		if n.OnNotified != nil {
-			// Notify the conflict detector background worker to assign the node to the worker asynchronously.
-			n.OnNotified(func() { n.assignTo(workerNum) })
-		} else {
-			// Assign the node to the worker directly.
-			n.assignTo(workerNum)
-		}
-	}
-}
-
-// tryResolve try to find a worker to assign the node to.
-// Returns (_, false) if there is a conflict,
-// returns (rand, true) if there is no conflict,
-// returns (N, true) if only worker N can be used.
-func (n *Node) tryResolve(resolvedDependencies, removedDependencies int32) (int64, bool) {
-	assignedTo, resolved := n.doResolve(resolvedDependencies, removedDependencies)
-	if resolved && assignedTo == assignedToAny {
-		assignedTo = n.RandWorkerID()
+func (n *Node) maybeReadyToRun() {
+	if ok := n.checkReadiness(); ok {
```

Comment:
When a large number of small transactions are in conflict, does performance degrade significantly?

Reply:
It is suppose not fast in the case of there are a large number of small transactions have conflict, no matter what you do.